transaction/core/src/tx_summary.rs

```diff
@@ -0,0 +1,220 @@
+//!   summary digest", and this is what the MLSAG actually sign.
+//!
+//! The TxSummary contains:
+//! * For each output in TxPrefix.outputs, the public key, target key, and
+//!   masked amount.
+//! * The list of pseudo output commitments.
+//!
+//! These data can be produced easily during Tx validation without major
+//! changes. They don't contain any secrets that the enclave isn't supposed to
+//! have.
+//!
+//! For the hardware wallet, what we can do is:
+//! * When we want to start signing MLSAGs, we send the 32 byte extended message
+//!   digest, and the TxSummary. The device can compute the "extended message
+//!   with tx summary digest" and sign MLSAGs over that.
+//! * For each output in the summary, the device expects the host computer to
+//!   ADDITONALLY supply this subset of the arguments to TxOut::new:
+//!   * block_version,
+//!   * amount,
+//!   * recipient,
+//!   * tx_private_key,
+//! * These are not part of the TxSummary (because the consensus enclave can't
+//!   know them) but can be part of TxSummaryUnblindingData or some such thing.
+//! * For a given public key, target key, and masked amount, it is intractable
+//!   to find a different amount, recipient, and tx private key that leads to
+//!   the same data (discrete log hard). So the device can be convinced that
+//!   these are the amounts and destinations of each such TxOut.
+//! * For each pseudo output commitment, either this corresponds to an MLSAG
+//!   that the device will actually sign, or it is coming from an SCI, which the
+//!   device will not actually sign. For anything which the device will not
+//!   actually sign, the TxSummaryUnblindingData can include the amount and
+//!   blinding factor. The computer actually has this for all of the pseudo
+//!   output commitments anyways.
+//!
+//! At this point, the device would then know the amount (value and token id) of
+//! every input in the Tx, and the amount and recipient of every output in the
+//! Tx. So it could completely account for the transfer of value caused by the
+//! Tx, it could identify change outputs, and it could display b58 encodings of
+//! the public addresses for outbound transfers. This should only require
+//! sending a few KB in the worst case.
+//!
+//! It's reasonable to ask, what if the host computer lies to the device in the
+//! TxSummary. What if, the TxPrefix actually says one thing, and the TxSummary
+//! it gives to the device says another. The device cannot detect this if it
+//! doesn't have insight into the extended message digest. However, in this
+//! case, the transaction will simply be invalid when it is submitted, because
+//! the consensus enclave will derive a different value for the TxSummary when
+//! it attempts to validate the transaction, and that value will actually be
+//! based on the TxPrefix. So the host computer can gain no advantage by lying
+//! to the device in this way.
```

Comment:
So this possibly constitutes a signing oracle in a way that could rely a bit more than would be ideal on the collision resistance of the transcript? I think it would depend on the implementation details of the hardware wallet, e.g. if you can just "plug it in and boop the 'OK' button" to sign, or if there is some PIN, etc., that is required. To be clear, this isn't broken or necessarily bad as-in, it's mostly just that I'd like to keep on the backburner to brainstorm potential ways to do it better.

Reply:
Yeah, I tried to think about this a bit i the MCIP, for me the issue is like:

* If the goal is that we don't send the whole Tx to the hardware wallet (to reduce bandwidth), then we have to accept that we might be signing something that isn't well-formed
* We could imagine restructuring the MLSAG signing digest so that it's like, first we digest the range proofs, merkle proofs, etc., then we digest the TxPrefix (which is reorganized to look more like the TxSummary), and that's the whole digest and there's no third step. But I'm not sure if this gets you away from the issue of, if there is a collision in merlin then you can be tricked about the first part of what you are signing (which was abbreviated to a hash)

Maybe I'm oversimplifying, but that's my justification anyways for why this state is okay. But it probably simplifies the situation if we move towards replacing the TxPrefix with the TxSummary in the Tx